File: app/src/main/java/org/tasks/dialogs/ImportTasksDialog.java
package org.tasks.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.res.Resources;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import androidx.fragment.app.DialogFragment;
import com.todoroo.andlib.utility.DialogUtilities;
import com.todoroo.astrid.backup.TasksXmlImporter;
import dagger.hilt.android.AndroidEntryPoint;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.backup.TasksJsonImporter;
import org.tasks.backup.TasksJsonImporter.ImportResult;
import org.tasks.ui.Toaster;

@AndroidEntryPoint
public class ImportTasksDialog extends DialogFragment {

  private static final String EXTRA_URI = "extra_uri";
  private static final String EXTRA_EXTENSION = "extra_extension";

  @Inject TasksXmlImporter xmlImporter;
  @Inject TasksJsonImporter jsonImporter;
  @Inject DialogBuilder dialogBuilder;
  @Inject Activity context;
  @Inject Toaster toaster;

  public static ImportTasksDialog newImportTasksDialog(Uri data, String extension) {
    ImportTasksDialog importTasksDialog = new ImportTasksDialog();
    Bundle args = new Bundle();
    args.putParcelable(EXTRA_URI, data);
    args.putString(EXTRA_EXTENSION, extension);
    importTasksDialog.setArguments(args);
    return importTasksDialog;
  }

  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    Bundle arguments = getArguments();
    Uri data = arguments.getParcelable(EXTRA_URI);
    String extension = arguments.getString(EXTRA_EXTENSION);
    ProgressDialog progressDialog = dialogBuilder.newProgressDialog();
    progressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
    progressDialog.setCancelable(false);
    progressDialog.setIndeterminate(true);
    progressDialog.show();
    setCancelable(false);
    switch (extension) {
      case "json":
        Handler handler = new Handler();
        new Thread(
                () -> {
                  ImportResult result =
                      jsonImporter.importTasks(getActivity(), data, progressDialog);
                  handler.post(() -> {
                    if (progressDialog.isShowing()) {
                      DialogUtilities.dismissDialog((Activity) context, progressDialog);
                    }
                    showSummary(result);
                  });
                })
            .start();
        break;
      case "xml":
        xmlImporter.importTasks(getActivity(), data, progressDialog);
        break;
      default:
        throw new RuntimeException("Invalid extension: " + extension);
    }
    return progressDialog;
  }

  private void showSummary(ImportResult result) {
    Resources r = context.getResources();
    dialogBuilder
        .newDialog(R.string.import_summary_title)
        .setMessage(
            context.getString(
                R.string.import_summary_message,
                "",
                r.getQuantityString(R.plurals.Ntasks, result.getTaskCount(), result.getTaskCount()),
                r.getQuantityString(
                    R.plurals.Ntasks,
                    result.getImportCount(),
                    result.getImportCount()),
                r.getQuantityString(R.plurals.Ntasks, result.getSkipCount(), result.getSkipCount()),
                r.getQuantityString(R.plurals.Ntasks, 0, 0)))
        .setPositiveButton(android.R.string.ok, (dialog, id) -> dialog.dismiss())
        .show();
  }
}

File: app/src/main/java/org/tasks/dialogs/ImportTasksDialog.kt
package org.tasks.dialogs

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import androidx.core.net.toUri
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.compose.ImportTasksViewModel
import org.tasks.themes.TasksTheme
import org.tasks.themes.Theme
import javax.inject.Inject

@AndroidEntryPoint
class ImportTasksDialog : DialogFragment() {

    @Inject lateinit var theme: Theme

    private val viewModel: ImportTasksViewModel by activityViewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NO_FRAME, android.R.style.Theme_Translucent_NoTitleBar)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val uri = requireArguments().getString(ARG_URI)!!.toUri()

        return ComposeView(requireContext()).apply {
            setContent {
                TasksTheme(
                    theme = theme.themeBase.index,
                    primary = theme.themeColor.primaryColor,
                ) {
                    org.tasks.compose.ImportTasksDialog(
                        uri = uri,
                        viewModel = viewModel,
                        onFinished = { dismiss() }
                    )
                }
            }
        }
    }

    companion object {
        private const val ARG_URI = "uri"

        fun newImportTasksDialog(uri: Uri): ImportTasksDialog {
            return ImportTasksDialog().apply {
                arguments = Bundle().apply {
                    putString(ARG_URI, uri.toString())
                }
            }
        }
    }
}
